pkg/internal/codegen/generator.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package codegen
+
+import (
+	"fmt"
+
+	"k8s.io/gengo/args"
+	"k8s.io/gengo/generator"
+	"k8s.io/gengo/namer"
+	"k8s.io/gengo/parser"
+)
+
+// Generator defines a generic generator for api parsing, code gen
+// (TODO): Design an upfront generic Generator that on top of gengo
+// for all parsing-api/code-gen relating works in controller-tool
+type Generator interface {
+	Inject(i interface{}) error
+	AddParam(i interface{})
+	NewContext() (*generator.Context, error)
+}
+
+// Param represents arguments or options that can be added to Generator
+type Param interface {
+	Add(i Generator)
+}
+
+type defaultGenerator struct {
+	*parser.Builder
+	*args.GeneratorArgs
+}
+
+var _ Generator = &defaultGenerator{}
+
+// DefaultGenerator provides a default API generator for parsing APIs and generating codes.
+func DefaultGenerator() (Generator, error) {
+	arguments := args.Default()
+	builder, err := arguments.NewBuilder()
+	return &defaultGenerator{builder, arguments}, err
+}
+
+func (a *defaultGenerator) Inject(i interface{}) error {
+	b, ok := i.(*parser.Builder)
+	if ok {
+		a.Builder = b
+		return nil
+	}
+	if arg, ok := i.(*args.GeneratorArgs); ok {
+		a.GeneratorArgs = arg
+		return nil
+	}
+	return fmt.Errorf("can not inject type (%T) in apiGenAdaptor", i)
+}
+
+// NewContext returns a default context based on generator.
+// (TODO): Applying this method, codegen/parse/context.go can be deprecated
+func (a *defaultGenerator) NewContext() (*generator.Context, error) {
+	return generator.NewContext(a.Builder, NameSystems(), DefaultNameSystem())
+}
+
+// AddParam adds input options to generator by its Add strategy.
+func (a *defaultGenerator) AddParam(i interface{}) {
+	p, ok := i.(Param)
+	if ok {
+		p.Add(a)
+	}
+}
+
+// DefaultNameSystem returns public by default.
+func DefaultNameSystem() string {
+	return "public"
+}
+
+// NameSystems returns the name system used by the generators in this package.
+// e.g. black-magic
+func NameSystems() namer.NameSystems {
+	return namer.NameSystems{
+		"public": namer.NewPublicNamer(1),
+		"raw":    namer.NewRawNamer("", nil),
+	}
+}
```

Comment:
Can we move this in a separate PR ? preferrably with use-case in existing code to demonstrate the utility ?

Reply:
Sure. Thanks